src/MagnumPlugins/KtxImageConverter/KtxImageConverter.cpp

```diff
@@ -0,0 +1,608 @@
+    /* Block header */
+    Implementation::KdfBasicBlockHeader& header = *reinterpret_cast<Implementation::KdfBasicBlockHeader*>(data.suffix(offset).data());
+    offset += sizeof(header);
+
+    header.vendorId = Implementation::KdfBasicBlockHeader::VendorId::Khronos;
+    header.descriptorType = Implementation::KdfBasicBlockHeader::DescriptorType::Basic;
+    header.versionNumber = Implementation::KdfBasicBlockHeader::VersionNumber::Kdf1_3;
+    header.descriptorBlockSize = dfdBlockSize;
+
+    header.colorModel = Implementation::KdfBasicBlockHeader::ColorModel::Rgbsda;
+    header.colorPrimaries = Implementation::KdfBasicBlockHeader::ColorPrimaries::Srgb;
+    header.transferFunction = isSrgb(format)
+        ? Implementation::KdfBasicBlockHeader::TransferFunction::Srgb
+        : Implementation::KdfBasicBlockHeader::TransferFunction::Linear;
+    /** @todo Do we ever have premultiplied alpha? */
+    header.bytesPlane[0] = texelSize;
+
+    /* Color channels */
+    const auto samples = Containers::arrayCast<Implementation::KdfBasicBlockSample>(data.suffix(offset));
+    offset += dfdSamplesSize;
+
+    const UnsignedByte bitLength = typeSize*8;
+
+    static const Implementation::KdfBasicBlockSample::ChannelId channelIdsRgba[4]{
+        Implementation::KdfBasicBlockSample::ChannelId::Red,
+        Implementation::KdfBasicBlockSample::ChannelId::Green,
+        Implementation::KdfBasicBlockSample::ChannelId::Blue,
+        Implementation::KdfBasicBlockSample::ChannelId::Alpha
+    };
+    /*
+    static const Implementation::KdfBasicBlockSample::ChannelId channelIdsDepthStencil[2]{
+        Implementation::KdfBasicBlockSample::ChannelId::Depth,
+        Implementation::KdfBasicBlockSample::ChannelId::Stencil
+    };
+    static const Implementation::KdfBasicBlockSample::ChannelId channelIdsStencil[1]{
+        Implementation::KdfBasicBlockSample::ChannelId::Stencil
+    };
+    */
+
+    /** @todo detect depth/stencil */
+    Containers::ArrayView<const Implementation::KdfBasicBlockSample::ChannelId> channelIds =
+        Containers::arrayView(channelIdsRgba);
+
+    UnsignedShort bitOffset = 0;
+    for(UnsignedInt i = 0; i != numChannels; ++i) {
+        auto& sample = samples[i];
+        sample.bitOffset = bitOffset;
+        sample.bitLength = bitLength - 1;
+        const auto channelId = channelIds[i];
+        sample.channelType = channelId | channelFormat(suffix, channelId);
+        const auto mapping = channelMapping(suffix, typeSize);
+        sample.lower = mapping[0];
+        sample.upper = mapping[1];
+
+        bitOffset += bitLength;
+
+        Utility::Endianness::littleEndianInPlace(sample.bitOffset,
+            sample.lower, sample.upper);
+    }
+
+    Utility::Endianness::littleEndianInPlace(length);
+    Utility::Endianness::littleEndianInPlace(header.vendorId, header.descriptorType,
+        header.versionNumber, header.descriptorBlockSize);
+
+    CORRADE_INTERNAL_ASSERT(offset == dfdSize);
+
+    return data;
+}
```

Comment:
Crazy idea (sorry in advance) -- I was wondering, instead of all the above code which seems rather involved and (to me at least) not trivial to extend for new formats, would it be practical to just store a huge table of pre-filled DFDs for each format we have, generated from the JSON for example? Assuming 256 formats (currently there's about 160) and (upper-bound) a 64 byte DFD for each (BC5 has 60), this would be 16 kB of data, which isn't too much I'd say. Then adding a new format would be just about regenerating from the JSON, and putting the DFD into the output would be just a copy from this array.

Apologies for the time spent on the code :sweat_smile:

Reply:
[formats.json](https://github.com/KhronosGroup/KTX-Specification/blob/master/formats.json) sadly won't help since it doesn't contain any info that Magnum doesn't already expose for existing `(Compressed)?PixelFormat`. I wanted to use that to get that info for arbitrary Vulkan formats, but for generating the DFD it's of no use.

One alternative would be using (parts of) the [Khronos DFD lib](https://github.com/KhronosGroup/KTX-Software/tree/master/lib/dfdutils), it looks self-contained and not terribly bloated.

Reply:
Mmmm, C code generated by *Perl* scripts. Nope nope nope, that sounds like a *much bigger* maintenance hell than I'd want to suffer. Ah well, I thought they would have some *data* file enumerating all DFDs for all Vulkan types, but if I have to run code in order to (again) get a C code and not plain data, that's not the way I want to go.

Keep it like this, then.